### ISSUE TITLE
Mocking/stubbing an ApplicationHelper method in a view spec.

### ISSUE THREAD

**Issue by jdickey** [rspec/rspec-rails#780]
There should be an easier/more intuitive way to double (stub or mock) `ApplicationHelper` methods called by view code from within a view spec (RSpec2, MiniTest, whatever).

Given a view (in HAML):

``` haml

%table.table.table-striped.table-bordered.table-hover
  %caption
    %h1 All registered users
  %thead
    %tr
      %th Pen Name
      %th Bio
  %tbody
  - @users.each do |user|
    %tr
      %td= user[:name]
      %td.bio!= format_newlines_as_html_paras user[:bio]
```

what I want to do in my view spec (RSpec2 here):

``` ruby

require 'spec_helper'

describe 'users/index' do

  before :each do
    @users = []
    3.times.each {@users << FactoryGirl.build_stubbed(:user)}
  end

  # ...

  describe 'calls the expected helpers, including' do

    it 'format_newlines_as_html_paras, once' do
      ActionView::Base.any_instance.stub(:format_newlines_as_html_paras).
        and_return "FOODOOFOO"
      render
      rendered.should have_content 'FOODOOFOO'
    end

  end # describe 'calls the expected helpers, including'

end # describe 'users/index'
```

That spec, as written, does not pass, and I've been flailing around for several hours trying to find a way that does.

What's the point?

By doubling the `ApplicationHelper` method within the view spec, we remove the dependency on the _actual_ `ApplicationHelper` method in order for the view spec to pass. We can use expectations to verify that the helper method is called the expected number of times, and we _don't care_ about the method's inputs or outputs at this point; we're asserting the _interface_, not the _implementation logic_ per se.

Any ideas?

**Comment by jrochkind** [comment]
Trying to stub a helper for a view test with:

   allow(view).to receive(:staff_user?).and_return(false)

As suggested above, and in a StackOverflow, results in:

~~~
#<RSpec::Mocks::MockExpectationError: #<#<Class:0x007fd62cbaaa90>:0x007fd6312cc7c0 @_routes=nil, [lots of stuff] >> does not implement: staff_user?>
~~~

Surely there's some way to provide mocked helper methods for views?

**Comment by JonRowe** [comment]
You can wrap your view stub in a `without_partial_double_verification { ... }` call or turn off partial double verification completely.

**Comment by jrochkind** [comment]
Thanks @JonRowe! I'll try that. I forgot about the "partial double verification" thing, didn't realize that's what was going on, or how to get around it. I'm surprised it was so hard to google for a solution for this, I'd think it would be a common need. If it works, I'll blog it and try to get another google hit at least, but ideally it might be in the rspec view spec docs? (which I guess means relish? Finding what you need in the relish docs is challenging, but that's another topic). 

Or are view specs going to go away entirely with Rails dropping them in 5.x? I hope not, I find them very useful.

**Comment by JonRowe** [comment]
> Or are view specs going to go away entirely with Rails dropping them in 5.x?

I wasn't aware Rails was dropping view tests, if they have they will also go from RSpec in those versions as RSpec Rails is closely pinned to Rails.

**Comment by jrochkind** [comment]
I may have been confused now I can't track down what i was thinking of.

**Comment by jrochkind** [comment]
@JonRowe I discovered I need 3.6.0.beta2 to use "without_partial_double_verification".  I'd rather not run pre-final software on my production app, although I guess for rspec that doesn't actually run in production perhaps I just should.  But there are other dependencies that seem to complain when I try at present. 

Is there any good way for me to temporarily turn off this config then turn it back on myself 'manually' in 3.5.0?

**Comment by JonRowe** [comment]
Hmm, guess we've been slacking on releases, 3.6.0 will be released soon-ish, (probably around the Rails 5.1 release) but to be honest our beta versions are usually production ready
